[Lee Kay Yen] iP

kayyenl opened the pull request:
## A new rendition of Duke: Muse!

How is this an improvement over Duke?
- More Functionality
- Memory Functionality
- ~~**More verbose**~~ More Comprehensive in Reply

Top 3 ways we recommend using this: 
1. As a shopping list :writing_hand:
2. To show off your coding skills :brain:
3. .....idk :ok_man:

How successful we have been with this Muse thing: 
- [x] Appeared on The Straight Times
- [ ] Appeared on some random guy's computer
- [x] Appeared on some guy's github
- [ ] Got my mom to use it

So, this project has gone places! If you feel discouraged seeing this, remember: 
> No code was built in a day. -Gandhi

We started with this small codebase:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

and we made it into this!
```java
public class DUKE {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

So, the [sky's](https://img.freepik.com/free-photo/white-cloud-sky-background_74190-4646.jpg?w=2000) the limit!

Commit:
Add Gradle support

kayyenl committed:
Level 1 done.

kayyenl committed:
finished level 2

kayyenl committed:
repaired some bugs to make level 2 look more proper

kayyenl committed:
finally done with level 3, learnt to use equals method

kayyenl committed:
still working on level 4, trying to format the words properly

kayyenl committed:
Finished with Level 4

kayyenl committed:
successfully ran the runtest.bat, and was able to get correct output

kayyenl committed:
Added the delete functionality, and separated child classes of Task to .java files.

kayyenl committed:
added an exception class.

kayyenl committed:
Managed to implement exceptions, learn to use throwing. Have OutOfBounds and BlankInput right now.

kayyenl committed:
updated the name of my exceptions to be more accurate.

kayyenl committed:
Create file for TaskList, so it is a separate class from the main

kayyenl committed:
Reorganise and allow for very basic file reading

kayyenl committed:
Implement the exceptions for marking and unmarking, when bool of task does not match

kayyenl committed:
Made loading functionality for save files, still not finishing level 7

kayyenl committed:
Implement Save, Load functionality and compatible with all current features

kayyenl committed:
Divide classes into individual classes (done during level 7)

kayyenl committed:
Dummy Commit before switching back to master branch

kayyenl committed:
Add exclaimation mark to the greeting (actually testing out branching ability)

kayyenl committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

kayyenl committed:
Done with Level 7

kayyenl committed:
Fix indexing error in the duke.java

kayyenl committed:
Implement date and time for Deadline class

kayyenl committed:
Handle Deadline already, managed to convert to string date

kayyenl committed:
Improved support for date when entering entries of Deadline class

kayyenl committed:
Implement Date and Time functionalities for Deadline and Event finally

kayyenl committed:
Forgot to merge the branch level 8 when doing, shall do now

kayyenl committed:
Try again, merge the commit file from branch level 8

kayyenl committed:
Commit from master to merge with branch level 8

kayyenl committed:
Test commit in master

kayyenl committed:
Remove one line from duke.txt

kayyenl committed:
Merge branch 'branch-Level-8'

kayyenl committed:
Merge remote-tracking branch 'origin/add-gradle-support'

kayyenl committed:
Edit some of the duke, trivial

kayyenl committed:
Add support for Gradle, now main() is run on Gradle

kayyenl committed:
Package exceptions into its own respective package (called duke.exceptions) and refactor

kayyenl committed:
Refactor exceptions into a separate package (duke.exceptions)

kayyenl committed:
Implemented User Input handling functionality inside Ui class, successfully debug lag problem in first input in Muse's program (always lags upon first input)

kayyenl committed:
Learn "throws IOException" in the method definition and applied it, it basically says that the exception will be handled by the class calling it one level above / levels above

kayyenl committed:
Add Storage, Ui, Parser and TaskList classes appropriately

kayyenl committed:
Fix another bug (when provide no extra argument for keyword, crash)

kayyenl committed:
Fix another bug (when u add a todo, once u mark, cannot unmark) this bug led to the wrong indexing of unmarked objects.

kayyenl committed:
Package classes into smaller packages

kayyenl committed:
A-JUnit Done, implemented 3 tests (2 methods each) for the 3 subclasses of Task (Todo, Event and Deadline)

kayyenl committed:
Create JAR file and released it on github, tested

kayyenl committed:
Annotate TaskList, Duke and Ui Classes

kayyenl committed:
Write documentations for all the helper classes' public functions

kayyenl committed:
Implement Find function, fixed more bugs (excessive white spaces on event and deadline's toString strings)

kayyenl committed:
Forgot to work on Level 9 in this branch... will push an empty commit now onto this branch

kayyenl committed:
second blank commit

kayyenl committed:
Do a commit on master so i can merge level 9 branch and master

kayyenl committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/helpers/Parser.java

kayyenl committed:
Done documentation for Event Class

kayyenl committed:
Document for Task Class (just the header)

kayyenl committed:
Merge branch 'branch-A-JavaDoc'

kayyenl committed:
Write docs for Task class

kayyenl committed:
Enforced checkstyle settings in ide, wraps at 120 words and spaces after operations

kayyenl committed:
About to merge with branch-A-CodingStyle

kayyenl committed:
Merge branch 'branch-A-CodingStandard'

kayyenl committed:
Remove irrelevant import statements in Ui

kayyenl committed:
Add checkstyle into gradle, runs well

Lava-Iris commented on duke.txt:
```diff
@@ -0,0 +1,6 @@
+[T] [X] work
```
If this is the task storage, it might be better to not track this using git as it is ever changing and is more of an output than a part of the code. You can get git to ignore this file by adding it to .gitignore and deleting this file.

kayyenl replied:
Thanks for the review & feedbacks! Agreed, I don't think I have been utilizing some of Git's features enough 😆

Lava-Iris commented on src/main/java/duke/helpers/Parser.java:
```diff
@@ -0,0 +1,117 @@
+            String[] splitArr = input.split(" ");
+            try {
+                if (input.equals("list")) {
+                    System.out.println(Ui.formatStr(tasks.listThings()));
```
Maybe switch can be used instead of if else here? It would increase readability and i think it's faster.

Lava-Iris commented on src/main/java/duke/helpers/Parser.java:
```diff
@@ -0,0 +1,117 @@
+                                "The number has to be within range of our task-list!\n" +
+                                "try again.");
+                    }
+                    if (splitArr[0].equals("mark") &&
```
I like that you handle the case where the user tries to mark something which is already done again

Lava-Iris commented on src/main/java/duke/helpers/TaskList.java:
```diff
@@ -0,0 +1,135 @@
+            this.addTask(newTodo);
+        }
+        if (line.substring(1,2).equals("D")) {
+            String deadline = line.substring(8);
```
maybe else if can be used here instead of if? Otherwise the code will check for D and E even when it knows it's a todo task.

Lava-Iris commented on src/main/java/duke/helpers/TaskList.java:
```diff
@@ -0,0 +1,135 @@
+            Deadline newDead = new Deadline(deadline, this.checkMark(line.substring(5,6)));
+            this.addTask(newDead);
+        }
+        if (line.substring(1,2).equals("E")) {
```
I feel there are some unnecessary repetitions in this method. I like to structure it like this: 
`String str = line.substring(8);` and `Task newTask;`  declared before the if statement. A value to newTask (like `newTask = new Todo(todo, this.checkMark(line.substring(5,6)));` in the if else statement. Then at the end `addTask(task);` after the if statement closes.

Lava-Iris reviewed (commented):
Hi!

The code is well structured and I couldn't find any major problems. 

All the best for future increments!

(P.S. I really like your bot's personality)

nicljr commented on src/main/java/duke/helpers/Parser.java:
```diff
@@ -0,0 +1,117 @@
+     * @param tasks Stores the Tasks objects into
+     * @param pw Writes to the screen for the user to comprehend.
+     */
+    public static void handleInputs(Scanner sc, TaskList tasks, PrintWriter pw) {
```
I like the naming of this function as it is able to convey its use!

nicljr commented on src/main/java/duke/helpers/Parser.java:
```diff
@@ -0,0 +1,117 @@
+                    throw new VagueInputException("Oh no! What do you mean? \n" +
+                            "I'm confused. Please specify... @.@");
+                }
+            } catch (VagueInputException ex) {
```
I like how you the namings of the different Exceptions as I am able to understand clearly what possible errors handling inputs will result in!

nicljr commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,44 @@
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        try {
+            Ui.doGreeting();
```
The naming of the various variables and methods do follow the coding standards and I like how all the methods start with verbs as it expresses clearly what it is trying to do

kayyenl replied:
Thank you for the generous feedbacks! I agree, naming conventions really help to give a nicer touch to any piece of code. All the best in your future increments! 😄

nicljr reviewed (commented):
Overall, I like the code structure and most, if not all, do follow the coding standards! I learnt a lot from your code as well! Thanks!

kayyenl committed:
Dummy Commit

kayyenl committed:
Make the default GUI for muse

kayyenl committed:
Implement functionality for GUI, have yet to include actual output by Muse

kayyenl committed:
Done somewhat barebones GUI of muse.

Added padding to the text boxes beside the profiles of user and muse.
Fixed several bugs, have not managed to close the app.

kayyenl committed:
Refactor code such that the saving functionality is still intact and working

kayyenl committed:
Implemented bye functionality

kayyenl committed:
commit before going into code quality branch

kayyenl committed:
Reduce level of nesting for certain deep nested statements in Parser, made sure no nested loop is beyond 2 nestings

kayyenl committed:
refactor Parser (most messy) to become more readable, and also showed the Happy Path

kayyenl committed:
Create GuiCustomiser, handles the visual intricate package functions that affect visuals of components.

By reducing the number of imperative lines in main, trying to make the main class more SLAP adherent.

kayyenl committed:
Refactor more of duke.start code, SLAP methodology to remove all deeper abstraction layer code

kayyenl committed:
Commit

kayyenl committed:
Add C-Help to provide helpsheet for new users

kayyenl committed:
Implement assert check to make sure commands are properly processed @ parser.java, using gradle to run

kayyenl committed:
Merge pull request #1 from kayyenl/branch-Assertions

PR practice + Assertion done

kayyenl committed:
no message

kayyenl committed:
Merge branch 'master' of https://github.com/kayyenl/ip

kayyenl committed:
Test commit to see if my pc and github are in sync

kayyenl committed:
begin commit this week, going to submit this proj after a few more change

kayyenl committed:
Add documentation for some of the visual classes (dialogbox)

kayyenl committed:
manage to get back the intellij workflow

kayyenl committed:
clean up code, make duke method shorter

kayyenl committed:
Improve quality of code (refactoring of parser method <30 LoC)

kayyenl committed:
reformat help sheet more correctly

kayyenl committed:
Add Ui.png

kayyenl committed:
Added more case handling

kayyenl committed:
Update README.md

Modify README for easier user usage

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md

kayyenl committed:
Update README.md
